src/sentry/utils/arroyo_guard.py

```diff
@@ -0,0 +1,112 @@
+import logging
+from abc import ABC, abstractmethod
+from collections import deque
+from collections.abc import Callable
+from typing import Any, Deque, Generic, TypeVar
+
+from arroyo.dlq import InvalidMessage
+from arroyo.processing.strategies import MessageRejected
+from arroyo.processing.strategies.abstract import ProcessingStrategy
+from arroyo.types import Message, TStrategyPayload
+
+logger = logging.getLogger(__name__)
+
+TResult = TypeVar("TResult")
+TResult_co = TypeVar("TResult_co", covariant=True)
+
+
+class HasNextStep(ABC, Generic[TResult]):
+    @abstractmethod
+    def __init__(self, next_step: ProcessingStrategy[TResult], *args: Any, **kwargs: Any) -> None:
+        pass
+
+
+class NonFinalStrategy(ProcessingStrategy[TStrategyPayload], HasNextStep[TResult]):
+    pass
+
+
+def guard(
+    max_buffer_size: int = 10,
+) -> Callable[
+    [type[NonFinalStrategy[TStrategyPayload, TResult]]],
+    type[NonFinalStrategy[TStrategyPayload, TResult]],
+]:
+    """
+    Handles backpressure and invalid messages for any strategy. If
+    MessageRejected or InvalidMessage is received from the inner strategy
+    from the inner strategy the message is carried over by the guard for
+    future submission / reraise so the parent strategy does not have to implement
+    any handling for this scenario and can safely call `next_step.submit()`
+    in any part of the strategy.
+
+    `next_step` must be the first arg in the constructor.
+
+    """
+
+    def guard_inner(
+        cls: type[NonFinalStrategy[TStrategyPayload, TResult]]
+    ) -> type[NonFinalStrategy[TStrategyPayload, TResult]]:
+
+        class Guard(NonFinalStrategy[TStrategyPayload, TResult]):
+            def __init__(
```

Comment:
this is a bit confusing, and I had to do a double-take to see what is being wrapped. since it appears to only wrap next_step, I think it can be modelled as a strategy around next_step. instead of:

```
@guard
class StepOne(...): ...
```

one could write:

```
class StepOne(...):
    def __init__(self, next_step):
        self.next_step = Guard(next_step)
```

the code in `poll()` can live in `Guard.poll`, as `StepOne` will call `Guard.poll()` anyway.

Reply:
oh.. this sounds way simpler!

Reply:
ok there's one thing i can't figure out how to deal with if we move this way

StepOne doesn't know when the guard's buffer is full, which is needed to raise MessageRejected to the parent

In the current implementation the guard controls both the parent and inner strategy and can do this

Do you have any ideas to deal with this?